ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -474,36 +484,90 @@ public Void visitFunctionCall(
-        final SqlType resolvedArgType = childContext.getSqlType();
+      final List<Expression> arguments = node.getArguments();
 
+      for (final Expression expression : arguments) {
         if (expression instanceof LambdaFunctionCall) {
-          argTypes.add(
+          argTypesForFirstPass.add(
               SqlArgument.of(
-                  SqlLambda.of(
-                      currentTypeContext.getLambdaInputTypes(),
-                      resolvedArgType)));
+                  SqlLambda.of(((LambdaFunctionCall) expression).getArguments().size())))
+          ;
         } else {
-          argTypes.add(SqlArgument.of(resolvedArgType));
-          // for lambdas - we save the type information to resolve the lambda generics
-          if (hasLambda) {
-            currentTypeContext.visitType(resolvedArgType);
-          }
+          final TypeContext childContext = expressionTypeContext.getCopy();
+          process(expression, childContext);
+          final SqlType resolvedArgType = childContext.getSqlType();
+          argTypesForFirstPass.add(SqlArgument.of(resolvedArgType));
         }
       }
 
-      final SqlType returnSchema = udfFactory.getFunction(argTypes).getReturnType(argTypes);
-      expressionTypeContext.setSqlType(returnSchema);
+      final KsqlScalarFunction function = udfFactory.getFunction(argTypesForFirstPass);
+
+      if (node.hasLambdaFunctionCallArguments()) {
+        final List<ParamType> paramTypes = function.parameters();
+        final Map<GenericType, SqlType> reservedGenerics = new HashMap<>();
+        final List<SqlArgument> argForReturnType = new ArrayList<>();
+        for (int i = 0; i < arguments.size(); i++) {
+          final Expression expression = arguments.get(i);
+          final ParamType parameter = paramTypes.get(i);
+          if (expression instanceof LambdaFunctionCall) {
+            final TypeContext childContext = expressionTypeContext.getCopy();
+
+            // the function returned from the UDF factory should have lambda
+            // at this index in the function arguments if there's a
+            // lambda node at this index in the function node argument list
+            if (!(parameter instanceof LambdaType)) {
+              throw new KsqlException("Error while processing lambda function. "
+                  + "This is most likely an internal error and a "
+                  + "Github issue should be filed for debugging.");
+            }
+
+            final LambdaType lambdaParameter = (LambdaType) parameter;
+            for (ParamType inputParam : lambdaParameter.inputTypes()) {
+              if (inputParam instanceof GenericType) {
+                final GenericType genericParam = (GenericType) inputParam;
+                if (!reservedGenerics.containsKey(genericParam)) {
+                  throw new KsqlException(
+                      String.format(
+                          "Could not verify type for generic %s.",
```

Comment:
also makes sense here to output the reserved generics map and what we expected (but didn't find)